File: _a_arm/matlab/hw7/armSim.m
armParamHW7;  % load parameters

% instantiate arm, controller, and reference input classes 
% Instantiate Dynamics class
alpha=0;
addpath('../hw3'); arm = armDynamics(alpha,P);
controller = armController(P);  
addpath('../hw2'); reference = signalGenerator(30*pi/180, 0.05); 
addpath('../hw2'); disturbance = signalGenerator(0.0, 0.0);

% instantiate the data plots and animation
addpath('../hw2'); dataPlot = dataPlotter(P);
addpath('../hw2'); animation = armAnimation(P);

% main simulation loop
t = P.t_start;  % time starts at t_start
y = arm.h();
while t < P.t_end  
    t_next_plot = t + P.t_plot;
    % Propagate dynamics in between plot samples
    while t < t_next_plot 
        r = reference.square(t);
        d = disturbance.step(t);
        x = arm.state;  % use the state and not the output
        u = controller.update(r, x);  % Calculate the control value
        y = arm.update(u+d);  % Propagate the dynamics
        t = t + P.Ts; % advance time by Ts
    end
    % update animation and data plots
    animation.update(arm.state);
    dataPlot.update(t, r, arm.state, u);
end

File: _a_arm/python/hw2/armSim.py
import matplotlib.pyplot as plt
import numpy as np
import sys
sys.path.append('..')  # add parent directory
import armParam as P
from signalGenerator import signalGenerator
from armAnimation import armAnimation
from dataPlotter import dataPlotter

# instantiate reference input classes
reference = signalGenerator(amplitude=0.5, frequency=0.1)
thetaRef = signalGenerator(amplitude=2.0*np.pi, frequency=0.1)
tauRef = signalGenerator(amplitude=5, frequency=.5)

# instantiate the simulation plots and animation
dataPlot = dataPlotter()
animation = armAnimation()

t = P.t_start  # time starts at t_start
while t < P.t_end:  # main simulation loop
    # set variables
    r = reference.square(t)
    theta = thetaRef.sin(t)
    tau = tauRef.sawtooth(t)
    # update animation

    state = np.array([[theta], [0.0]])
    animation.update(state)
    dataPlot.update(t, r, state, tau)

    #plt.show()
    t = t + P.t_plot  # advance time by t_plot
    plt.pause(0.1)

# Keeps the program from closing until the user presses a button.
print('Press key to close')
plt.waitforbuttonpress()
plt.close()
